### PR TITLE
cuda: allow user to set NVCC explicitly

### PR DESCRIPTION
**Pull request by spickartz**
## Pull Request Description
We cannot use icc as the host compiler since it conflicts with recent CUDA versions. Therefore, we rely on the default GCC as it is binary compatible with icc.

This is done by simply omitting to set the `-ccbin icc`.  The test in `subconfigure.m4` can also be successfully passed by adding a `-std=c++11`.  However, this still prevents yaksa from being build by using icc as the host compiler.

## Expected Impact
CUDA code will be compiled by using GCC as the host compiler if icc is used within the build. 

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Have read and agree to the Yaksa CLA terms (https://github.com/pmodels/yaksa/wiki/Yaksa-Contributor-License-Agreement)

### CODE REVIEW

File: src/backend/cuda/subconfigure.m4 [hunk 1]
```diff
@@ -62,7 +62,9 @@ if test "$with_cuda" != "no" ; then
     PAC_CHECK_HEADER_LIB([cuda_runtime_api.h],[cudart],[cudaStreamSynchronize],[have_cuda=yes],[have_cuda=no])
     if test "${have_cuda}" = "yes" ; then
         AC_MSG_CHECKING([whether nvcc works])
-        if test -n "$ac_save_CC" ; then
+        # only explicitly set the host compiler if not using the icc Intel
+        # compiler (this conflicts with current versions of nvcc)
+        if test -n "$ac_save_CC" -a "${ac_save_CC}" != "icc"; then
```
Comment (hzhou) [inline src/backend/cuda/subconfigure.m4:67]: Did you try `icpc` -- similar to what did for `pgcc` and `nvc`?

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4:67]: No, this does not work either. You can simply reproduce this by running:
```console
$ echo '__global__ void foo(int x) {}' | nvcc -x cu -c -ccbin icpc -
```

Instead the more recent `icx` and `icpx` compilers are working (given `-allow-unsupported-compiler` is set; which is probably something we do not want).

**Comment by spickartz** [comment]
I changed the logic to only try the fallback to the default host compiler if it fails with the specified `-ccbin` in the first place. This way `icc` would still be used if the version being tested is compatible with the installed CUDA version. What do you think @hzhou?

File: src/backend/cuda/subconfigure.m4 [hunk 2]
```diff
@@ -90,6 +92,29 @@ if test "$with_cuda" != "no" ; then
             have_cuda=no
```
Comment (hzhou) [inline src/backend/cuda/subconfigure.m4:92]: How about move the retest here?

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4:92]: Did you mean it this way? I now additionally set the host compiler explicitly to gcc instead of relying on the default.

File: src/backend/cuda/subconfigure.m4 [hunk 3]
```diff
@@ -89,7 +91,30 @@ if test "$with_cuda" != "no" ; then
         ],[
             have_cuda=no
             AC_MSG_RESULT([no])
+
+            # if using the Intel icc compiler retry without the '-ccbin' option and fallback to
+            # binary compatible gcc
+            if test "${ac_save_CC}" == "icc"; then
+                NVCC_FLAGS="$save_NVCC_FLAGS -ccbin gcc"
```
Comment (hzhou) [inline src/backend/cuda/subconfigure.m4:98]: I am thinking that we can just try the default (remove the `-ccbin` option) and see if it works. I would rather not to maintain the extra logic of checking `icc` and picking `gcc`.

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4:98]: Okay. I was just thinking that we can ensure binary compatibility  this way (with the C code of the non-CUDA files compiled using `icc`). Or don't you see an issue there?

Comment (hzhou) [reply src/backend/cuda/subconfigure.m4:98]: How about when the check failed here we simply spit a "helpful" message and ask users to set `NVCC` and re-configure? We did this in `test/mpi/configure` but neglected to do the same for yaksa -- https://github.com/pmodels/mpich/pull/5968

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4:98]: Mh aren't we having the exact the same issue [here](https://github.com/pmodels/mpich/blob/main/test/mpi/configure.ac#L736) and even omit the test during configure time? I would assume (did not try it out yet) that the compile would fail when `icc` is used as host compiler.

Probably it is a matter of taste. Do we wann to provide a fallback (and additionally print a warning in such cases). Or do we rather want the user to explicitly set the host compiler if the currently chosen compiler conflicts with `nvcc`?

Comment (hzhou) [reply src/backend/cuda/subconfigure.m4:98]: I think we should always have a mechanism to allow users to apply their own settings. So only do the check when `NVCC` is not set by the user. Since users can fallback on that, I think we can just fail for those corner cases such as `icc`.

Yes, we should test the `ccbin` at least once in configure.

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4]: Okay, I changed it accordingly. Users may now explicitly  set `NVCC` (including flags/options). The retry has been removed.

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4]: What do you think? Is this was you intended?

File: src/backend/cuda/subconfigure.m4 [hunk 4]
```diff
@@ -57,28 +57,36 @@ AC_ARG_WITH([cuda-sm],
...
         else
-            nvcc_bin=nvcc
+            nvcc_bin=$NVCC
+            AC_MSG_WARN([Using user-provided nvcc: '${nvcc_bin}'])
         fi
 
         # save language settings, customize ac_ext and ac_compile to support CUDA
         AC_LANG_PUSH([C])
         ac_ext=cu
         ac_compile='$nvcc_bin $NVCC_FLAGS -c conftest.$ac_ext >&5'
```
Comment (hzhou) [inline src/backend/cuda/subconfigure.m4:88]: Sorry for the late response. How about make it `ac_compile="$NVCC -c ...`, i.e. `NVCC="$nvcc_bin $NVCC_FLAGS"`? The intention is to allow user fully overwrite `NVCC`.

I am surprised to see single quotes here. Wouldn't the single quote prevent shell variable expansion? What did I miss?

Comment (hzhou) [reply src/backend/cuda/subconfigure.m4:88]: > > I am surprised to see single quotes here. Wouldn't the single quote prevent shell variable expansion? What did I miss?
> 
> Mh good point, oversaw that but it is working as expected, e.g., I saw something like `configure:18564: WARNING: Using user-provided nvcc: 'nvcc -ccbin gcc'` when testing. I did not take a closer look at the autoconf macros but would assume they expand, don't they?

That warning is from line 82, where the single quote is inside a double-quoted string.
Line 88 is not an autoconf macro. It's normal shell variables. Since it works, I am probably missed something here. I'll follow up.

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4:88]: > Line 88 is not an autoconf macro. It's normal shell variables. Since it works, I am probably missed something here. I'll follow up.

Yes, I know. But `ac_compile` is certainly used (and then expanded) by an autoconf macro I would have assumed.

Comment (spickartz) [reply src/backend/cuda/subconfigure.m4]: Fixed this and also changed the behavior w.r.t. `NVCC` and `NVCC_FLAGS`.

**Review by hzhou — APPROVED** [review]
Thanks for the changes!

**Comment by hzhou** [comment]
test:yaksa/gpu
